Me: Use Stripe Elements for "Add Card" form

sirbrillig opened the pull request:
#### Changes proposed in this Pull Request

This uses the Stripe Elements form fields introduced in #34469 for the "Add Card" form at `/me/purchases/add-credit-card`. Previously they were only used in the Checkout form (and only in some circumstances).

![Screen Shot 2019-08-02 at 6 10 25 PM](https://user-images.githubusercontent.com/2036909/62401418-13c27400-b551-11e9-84d3-9d7e4b6cd58b.png)

To do:

- [x] Split off fixes for CreditCardForm. #34963
- [x] Split off refactor of AddCreditCard to functional component. #34980
- [x] Split off moving hooks to lib/stripe. #34981
- [x] Fix form styles so that stripe elements look like regular elements (not sure why those style aren't being used).
- [x] Show validation errors on stripe fields.
- [x] Make form submission work with stripe fields 
- [x] Make sure form submission handles 3DS authentication.
- [x] Make sure payments made with these cards are successful.
- [x] Make sure `use_for_existing` works as expected (there's an issue with using it on cards added with stripe details).

## Testing instructions

- Sandbox the API and the store.
- Load Calypso with this PR using either a local copy or the calypso.live implementation.
- Make sure you have a purchase on your test site.
- On the front-end, visit `/me/purchases/add-credit-card`. Make sure that you see the Stripe form fields (placeholder numbers in the credit card number field, not dots).
- Enter a new test card with a unique Cardholder Name (you'll need to identify this name later) and click "Save Card".
- Verify that you get a success message in the upper right.
- Visit `/me/purchases` and click on any upgrade.
- In the lower right, click on the "Payment method" text/icon.
- Be sure the cardholder name that shows up matches the one you entered above.

Because this diff also touches the checkout flow, please also follow the testing instructions in #34469.

sirbrillig commented:
I did a bunch of refactoring to make it more obvious how the add-card form submits its data to the server (and remove a bunch of unused state). That can all go into a separate PR once this is ready.

We should be able to modify the form's flow to use `createStripePaymentMethod()` and then send the resulting Stripe Payment Method ID to `storedCardAdd()` via `saveStoredCard()` instead of the Paygate token created in `CreditCardForm.saveCreditCard()` and `createCardToken()`. Working on figuring how to do this...

sirbrillig commented:
Moved CreditCardForm refactor to #34963. It frees this code from the constraints of the (very complex) formStateController and should make it much easier to override the form field validation we will be using here.

sirbrillig commented:
D30582-code and D31015-code are necessary to test this currently.

sirbrillig commented:
Now requires D31044-code to request the setup intent.

sirbrillig commented:
@yoavf You managed to make the stripe elements use our form's styles in the last PR. Can you figure it out for this one? It's the same component (`CreditCardFormFields`) that's rendering them so I'm not sure why they would not be used.

sirbrillig commented:
There's some UX that might not be great here: if adding the card fails because of some obscure error on the server, the SetupIntent will already have been used, which can cause errors like `You cannot update this SetupIntent because it has already succeeded.` to be displayed to the user (since we catch all errors in that block). We might want to give a little more thought to the error flow here.

sirbrillig commented:
Requires D31104-code to make updating current subscriptions work.

sirbrillig commented:
The styles weren't showing because they were actually contained in the stylesheet for the Checkout component. I've moved them to the CreditCardForm component's stylesheet, but none of the styles in that file appear to be loading. I'm not familiar with Calypso's style loading system but there must be a bug here.

yoavf commented:
The time it takes for the stripe fields to be ready is a bit too long for my liking, now that we've added the SetupIntent creation. Let's disable the fields and add some kind of loader indication until that's ready? We can do that in a seperate PR.

![Screen Capture on 2019-08-04 at 11-27-03](https://user-images.githubusercontent.com/844866/62421477-eee21400-b6aa-11e9-9600-1699493af328.gif)

yoavf commented on client/me/purchases/add-credit-card/index.jsx:
```diff
@@ -21,21 +21,25 @@ import Main from 'components/main';
 	const goToBillingHistory = () => page( billingHistory );
 	const recordFormSubmitEvent = () =>
 		analytics.tracks.recordEvent( 'calypso_add_credit_card_form_submit' );
+	const shouldUseStripe = true; // TODO: how should we derive this?
```
When I force this to `false` and use the form, I get a `createCardToken is not a function` notice in the UI, and no action is taken.

sirbrillig replied:
Wow, thanks for seeing that! 😰 I mixed up my arguments in `makeAsyncCreateCardToken`

sirbrillig commented:
> The time it takes for the stripe fields to be ready is a bit too long for my liking, now that we've added the SetupIntent creation. Let's disable the fields and add some kind of loader indication until that's ready? 

Good point. That had been bothering me too. I added an `isStripeLoading` prop in the HOC which will be set when `stripe` is still undefined; that way the components can tell that they're loading stripe elements and show disabled fields. However, I can't render disabled _stripe elements_ fields, because the stripe elements fields won't render without stripe being loaded (catch-22), so I'm displaying the old-style fields and disabling _them_. That's not ideal since it's using the old-style code and the fields look different, but I'm not sure what a better UX would be at the moment. I also added the same feature to the checkout flow.

sirbrillig commented:
Extracted the Checkout and most of the tangential lib/stripe changes to #35196
